symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,311 @@
+                         unsigned int l) const
+{
+    std::vector<UIntDict> res;
+    size_t r = f_list.size();
+    auto lc = get_lc();
+
+    if (r == 1) {
+        integer_class g, r, s, pl;
+        mp_pow_ui(pl, p, l);
+        mp_gcdext(g, r, s, lc, pl);
+        UIntDict F((*this) * r);
+        F.itrunc(pl);
+        res.push_back(F);
+        return res;
+    }
+
+    std::vector<UIntDict> sub1, sub2;
+    integer_class m(p);
+    size_t k(r / 2);
+    auto d = std::ceil(std::log2(l)); // TODO
+    auto g = GaloisFieldDict::from_vec({lc}, p);
+
+    for (unsigned i = 0; i < k; ++i) {
+        g *= GaloisFieldDict(f_list[i].dict_, p);
+        sub1.push_back(f_list[i]);
+    }
+
+    sub2.push_back(f_list[k]);
+    auto h = GaloisFieldDict(f_list[k].dict_, p);
+    for (unsigned i = k + 1; i < f_list.size(); ++i) {
+        h *= GaloisFieldDict(f_list[i].dict_, p);
+        sub2.push_back(f_list[i]);
+    }
+
+    GaloisFieldDict temp2, s, t;
+    GaloisFieldDict::gf_gcdex(g, h, outArg(s), outArg(t), outArg(temp2));
+
+    UIntDict dg = UIntDict::from_vec(g.dict_);
+    dg.itrunc(p);
+    UIntDict dh = UIntDict::from_vec(h.dict_);
+    dh.itrunc(p);
+    UIntDict ds = UIntDict::from_vec(s.dict_);
+    ds.itrunc(p);
+    UIntDict dt = UIntDict::from_vec(t.dict_);
+    dt.itrunc(p);
+
+    for (unsigned int i = 1; i <= d; ++i) {
+        UIntDict::zz_hensel_step(m, (*this), dg, dh, ds, dt, outArg(dg),
+                                 outArg(dh), outArg(ds), outArg(dt));
+        mp_pow_ui(m, m, 2);
+    }
+
+    res = dg.zz_hensel_lift(p, sub1, l);
+    auto temp = dh.zz_hensel_lift(p, sub2, l);
+    res.insert(res.end(), temp.begin(), temp.end());
+    return res;
+}
+
+void UIntDict::zz_divide(const UIntDict &a, const UIntDict &b,
+                         const Ptr<UIntDict> &quo, const Ptr<UIntDict> &rem)
+{
+    auto b_poly = b;
+    if (a.size() == 0)
+        throw std::runtime_error("Zero Division Error");
+    map_uint_mpz res;
+    UIntDict tmp;
+    integer_class q, r;
+    unsigned int a_deg, b_deg;
+    while (b_poly.degree() >= a.degree()) {
+        a_deg = a.degree();
+        b_deg = b_poly.degree();
+        q = b_poly.get_lc() / a.get_lc();
+        res[b_deg - a_deg] = q;
+        tmp = UIntDict(map_uint_mpz{{b_deg - a_deg, q}});
+        b_poly -= (a * tmp);
+    }
+    *quo = UIntDict(res);
+    *rem = b;
+    *rem -= (a * (*quo));
+}
+
+std::set<RCP<const UIntPoly>, RCPBasicKeyLess> UIntPoly::zz_zassenhaus() const
+{
+    std::set<RCP<const UIntPoly>, RCPBasicKeyLess> factors;
+    unsigned int n = get_degree();
+    UIntDict f(this->poly_);
+    if (n == 1) {
+        factors.insert(rcp_from_this_cast<const UIntPoly>());
+        return factors;
+    }
+    integer_class fc(this->poly_.dict_.begin()->second);
+    auto A = mp_get_ui(this->poly_.max_abs_coef());
+    auto b = get_lc();
+    auto B
+        = std::abs(int(std::sqrt(n + 1)) * std::pow(2, n) * A * mp_get_si(b));
```

Comment:
It might not fit, but we have a bottleneck here.
In the algorithm we have:
```
    auto l = std::ceil(std::log(2 * B + 1) / std::log(mp_get_ui(fsqf.first)));
    integer_class pl;
    mp_pow_ui(pl, fsqf.first, l);
```
as we have to use `pl` as `modulo_` that means we need a similar sized `dict_`.
and `pl` is approximately equal to `2*B`.
`B = std::abs(int(std::sqrt(n + 1)) * std::pow(2, n) * A * mp_get_si(b));`
and `A` is already greater than `b`, so I think we can use a signed int here, because if it is greater than a signed int, `pl` would also be and we have a bottleneck due to vector's size.